src/main/kotlin/no/nav/syfo/client/tokendings/TokendingsClient.kt

```diff
@@ -0,0 +1,124 @@
+            .keyID(privateKey.keyID)
+            .type(JOSEObjectType.JWT)
+            .build()
+    }
+
+    suspend fun getOnBehalfOfToken(
+        scopeClientId: String,
+        token: String,
+    ): TokenendingsToken {
+        val cacheKey = token + scopeClientId
+        val cachedToken = tokenMap.get(cacheKey)
+        return if (cachedToken?.isExpired() == false) {
+            COUNT_CALL_TOKENDINGS_TOKEN_OBO_CACHE_HIT.increment()
+            cachedToken
+        } else {
+            val tokenendingsToken = getAccessToken(
+                Parameters.build {
+                    append("client_id", tokenxClientId)
+                    append("grant_type", "urn:ietf:params:oauth:grant-type:token-exchange")
+                    append("client_assertion_type", "urn:ietf:params:oauth:grant-type:jwt-bearer")
+                    append("client_assertion", getClientAssertion().serialize())
+                    append("subject_token_type", "urn:ietf:params:oauth:token-type:jwt")
+                    append("subject_token", token)
+                    append("audience", scopeClientId)
+                }
+            )?.toTokenendingsToken()
+                ?: throw RuntimeException("Failed to get obo-token from tokenendings")
+            tokenendingsToken.also {
+                COUNT_CALL_TOKENDINGS_TOKEN_OBO_CACHE_MISS.increment()
+                tokenMap.put(cacheKey, it)
+            }
+        }
+    }
+
+    private suspend fun getAccessToken(params: Parameters): TokenendingsTokenDTO? {
+        return try {
+            val response: HttpResponse = httpClient.post(tokendingsEndpoint) {
+                accept(ContentType.Application.Json)
+                setBody(FormDataContent(params))
+            }
+            response.body<TokenendingsTokenDTO>()
+        } catch (e: ClientRequestException) {
+            handleUnexpectedResponseException(e)
+            null
+        } catch (e: ServerResponseException) {
+            handleUnexpectedResponseException(e)
+            null
+        }
+    }
+
+    private fun handleUnexpectedResponseException(
+        responseException: ResponseException,
+    ) {
+        log.error(
+            "Error while requesting token from TokenDings with statusCode=${responseException.response.status.value}",
+            responseException
+        )
+    }
+
+    private fun getClientAssertion(): SignedJWT {
+        val fromTime = getNotBeforeTime()
+        val jwtClaimSet = JWTClaimsSet.Builder()
+            .audience(tokendingsEndpoint)
+            .subject(tokenxClientId)
+            .issuer(tokenxClientId)
+            .jwtID(UUID.randomUUID().toString())
+            .notBeforeTime(fromTime)
+            .expirationTime(getExpirationTime())
+            .issueTime(fromTime)
+            .build()
+        return SignedJWT(jwsHeader, jwtClaimSet).also {
+            it.sign(jwsSigner)
+        }
+    }
+
+    private fun getNotBeforeTime(): Date {
```

Comment:
Jeg skjønner ikke hva `not before time` betyr 🤔

Reply:
Se nesten nederst på denne siden: https://doc.nais.io/security/auth/tokenx/

Reply:
Takk! Jeg synes kanskje de kunne funnet et bedre navn der da 🙃